processing/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -758,4 +758,25 @@ private static Set<String> map(
       return results;
     }
   }
+
+  /**
+   * Returns a possibly less safe version of this expression.
+   *
+   * For top level expressions use {@link Expr#makeSingleThreaded(Expr)} to
+   * obtain the single-threaded version of the expression tree.
+   */
+  default Expr singleThreaded()
```

Comment:
This part is a usage mistake waiting to happen. People will call `expr.singleThreaded()` and then be surprised that it doesn't actually make a single-threaded expr tree. IMO it's better for this method to apply a visitor that goes and replaces all the ConstantExprs, even though it takes some control away from non-constant Exprs to customize their behavior.

Reply:
I've made changes to this and replaced it with a better approach